### PR TITLE
Fixes #17080

### LINKED ISSUE #17080: Optional destructured parameters do not work with default values
**TypeScript Version:**  2.4.0

**Code**

The following code incorrectly type checks:
```ts
interface Attrs {
    foo?: number;
    bar?: number;
}

function unsafe({ foo, bar }: Attrs = {}): number {
    return foo * bar.toString().length + 1;
}
```

**Expected behavior:**
This should fail to type check since `foo` and `bar` may both be of type `undefined`

**Actual behavior:**

Match behaviour of omitting default value:
```ts
function failsToTypecheck({ foo, bar }: Attrs): number {
    return foo * bar.toString().length + 1;
}
```
Or behaviour of omitting destructuring:
```ts
function failsToTypecheck(attrs: Attrs = {}): number {
    return attrs.foo * attrs.bar.toString().length + 1;
}
```

### PR DESCRIPTION
**Pull request by Lazarus535**
Changes are in src/compiler.checker.ts only
The second arguments to the function "removeOptionalityFromDeclaredType" has been changed from "getRootDeclaration(declaration)" to "declaration".

Fixes #

### CODE REVIEW

File: pull_request_template.md [hunk 1]
```diff
@@ -2,15 +2,16 @@
 Thank you for submitting a pull request!
```
Comment (sandersn) [inline pull_request_template.md]: I don’t think this file was supposed to be part of the PR.

File: tests/cases/compiler/optionalParameterInDestructuringWithInitializer.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+interface Foo {
+  readonly bar?: number;
+}
+
+function performFoo({ bar }: Foo = {}) {
+  useBar(bar);
+}
+
+function useBar(bar: number) {
```
Comment (sandersn) [inline tests/cases/compiler/optionalParameterInDestructuringWithInitializer.ts]: You can just `declare function useBar(bar: number): void;`

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -13046,7 +13046,7 @@ namespace ts {
                 node.parent.kind === SyntaxKind.NonNullExpression ||
                 declaration.kind === SyntaxKind.VariableDeclaration && (<VariableDeclaration>declaration).exclamationToken ||
                 declaration.flags & NodeFlags.Ambient;
-            const initialType = assumeInitialized ? (isParameter ? removeOptionalityFromDeclaredType(type, getRootDeclaration(declaration) as VariableLikeDeclaration) : type) :
+            const initialType = assumeInitialized ? (isParameter ? removeOptionalityFromDeclaredType(type, declaration as VariableLikeDeclaration) : type) :
```
Comment (sandersn) [inline src/compiler/checker.ts]: Any idea why getRootDeclaration was here? Was this copied from some other source that would also have an erroneous call to it?

**Comment by sandersn** [comment]
Approved after the two small fixes requested.

File: pull_request_template.md [hunk 4]
```diff
@@ -1,16 +0,0 @@
-<!--
```
Comment (sandersn) [inline pull_request_template.md]: Now this file is deleted, but shouldn't be.

Comment (Lazarus535) [reply pull_request_template.md]: Sry about that... :-(

**Comment by sandersn** [comment]
Thanks @Lazarus535 ! Our test coverage was pretty bad here and is now much better.

**Pull request merged**